include/appbase/application_base.hpp

```diff
@@ -0,0 +1,327 @@
+          * on first access.  This allows loose and deferred binding between plugins
+          *
+          * @tparam MethodDecl - @ref appbase::method_decl
+          * @return reference to the method described by the declaration
+          */
+         template<typename MethodDecl>
+         auto get_method() -> std::enable_if_t<is_method_decl<MethodDecl>::value, typename MethodDecl::method_type&>
+         {
+            using method_type = typename MethodDecl::method_type;
+            auto key = std::type_index(typeid(MethodDecl));
+            auto itr = methods.find(key);
+            if(itr != methods.end()) {
+               return *method_type::get_method(itr->second);
+            } else {
+               methods.emplace(std::make_pair(key, method_type::make_unique()));
+               return  *method_type::get_method(methods.at(key));
+            }
+         }
+
+         /**
+          * Fetch a reference to the channel declared by the passed in type.  This will construct the channel
+          * on first access.  This allows loose and deferred binding between plugins
+          *
+          * @tparam ChannelDecl - @ref appbase::channel_decl
+          * @return reference to the channel described by the declaration
+          */
+         template<typename ChannelDecl>
+         auto get_channel() -> std::enable_if_t<is_channel_decl<ChannelDecl>::value, typename ChannelDecl::channel_type&>
+         {
+            using channel_type = typename ChannelDecl::channel_type;
+            auto key = std::type_index(typeid(ChannelDecl));
+            auto itr = channels.find(key);
+            if(itr != channels.end()) {
+               return *channel_type::get_channel(itr->second);
+            } else {
+               channels.emplace(std::make_pair(key, channel_type::make_unique()));
+               return  *channel_type::get_channel(channels.at(key));
+            }
+         }
+
+         /**
+          * Post func to run on io_service with given priority.
+          *
+          * @param priority can be appbase::priority::* constants or any int, larger ints run first
+          * @param func function to run on io_service
+          * @return result of boost::asio::post
+          */
+         template <typename Executor, typename Func>
+         auto post( Executor &exec, int priority, Func&& func ) {
+            return exec.post(priority, std::forward<Func>(func));
+         }
+
+         const bpo::variables_map& get_options() const;
+
+         /**
+          * Set the current thread schedule priority to maximum.
+          * Works for pthreads.
+          */
+         void set_thread_priority_max();
+
+      void set_stop_executor_cb(std::function<void()> cb) {
+         stop_executor_cb = std::move(cb);
+      }
+
+      void set_post_cb(std::function<void (int, std::function<void()>)> cb) {
+         post_cb = std::move(cb);
+      }
```

Comment:
These two are not indented like the others.
I don't think either one of these methods is needed.

Reply:
I will fix the indentation. These two methods are needed and used in the `application` constructor (application_instance.hpp:42`), as they indirectly provide access to the executor from inside `application_base` member functions (see `application_base.cpp`) who doesn't know about it as it is a member of `application.hpp`).

Reply:
OK, the new header files are now consistently indented.